### PR TITLE
[mtg-598] [mtg-576] Penalties/slashing + decrease rewards

### PR DESCRIPTION
**Pull request by kstepanovdev**

### CODE REVIEW

File: programs/mpl-staking/src/instructions/penalties/slash.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        let voter = &mut ctx.accounts.voter.load_mut()?;
+        // We should forbid slashes for votes that are in active votings state.
+        // Otherwise consequences are unpredictable.
+        // Not applicable for tokens that don't contribute to voting power.
+        let token_owner_record = load_token_owner_record(
+            &voter.voter_authority,
+            &ctx.accounts.token_owner_record.to_account_info(),
+            &registrar,
+        )?;
+        token_owner_record.assert_can_withdraw_governing_tokens()?;
```
Comment (StanChe) [inline programs/mpl-staking/src/instructions/penalties/slash.rs:119]: More as a mental exercise - wdyt of marking the slashed amount on the voter account and decreasing the rewards amounts and only transfer the slashed amount on withdrawal? This way we should keep the governance happy and be able to slash without any way to overcome it with voting.

File: programs/mpl-staking/src/instructions/stake.rs [hunk 2]
```diff
@@ -61,17 +61,17 @@ pub fn stake(
         .checked_add(amount)
         .ok_or(MplStakingError::ArithmeticOverflow)?;
     target.delegate_last_update_ts = clock_unix_timestamp();
+    target.lockup.end_ts = target
```
Comment (StanChe) [inline programs/mpl-staking/src/instructions/stake.rs:64]: we should also apply this change on v1 (the audited contract)

File: programs/mpl-staking/src/lib.rs [hunk 3]
```diff
@@ -103,6 +103,7 @@ pub mod mpl_staking {
     }
 
     pub fn withdraw(ctx: Context<Withdraw>, deposit_entry_index: u8, amount: u64) -> Result<()> {
+        // instructions::withdraw(ctx, deposit_entry_index, amount);
```
Comment (StanChe) [inline programs/mpl-staking/src/lib.rs:106]: a duplicate comment?

**Review by StanChe — APPROVED** [review]
🦸 👍

File: programs/mpl-staking/src/instructions/penalties/slash.rs (hunk 1 above)
Comment (kstepanovdev) [reply programs/mpl-staking/src/instructions/penalties/slash.rs:119]: That's a good idea. 
However, we need to extend an account to keep track of how many tokens have been slashed.

Comment (kstepanovdev) [reply programs/mpl-staking/src/instructions/penalties/slash.rs:119]: On the other hand, it means unless the user presses "withdraw", tokens will be on his stake forever. Probably, we don't want this. It's simpler to wait until the vote expires.

**Pull request merged**